src/profiler/profiler.c

```diff
@@ -158,15 +160,43 @@ void xdebug_profile_call_entry_dtor(void *dummy, void *elem)
 	xdfree(ce);
 }
 
-static void profiler_write_header(FILE *file, char *script_name)
+static void profiler_fflush(char incl_file_flush)
+{
+	fwrite(XG_PROF(profile_file_buffer), sizeof(char), XG_PROF(profile_file_buffer_len), XG_PROF(profile_file));
+	XG_PROF(profile_file_buffer_len) = 0;
+
+	if (incl_file_flush) {
+		fflush(XG_PROF(profile_file));
+	}
+}
+
+static inline void profiler_fprintf_pre()
+{
+	if (XG_PROF(profile_file_buffer_len) >= PROFILE_FILE_BUFFER_SIZE / 2) {
+		profiler_fflush(0);
+	}
+}
+
+static inline void profiler_fprintf_post(int len)
+{
+	if (len >= 0) {
+		XG_PROF(profile_file_buffer_len) += len;
+	}
+}
+
+#define profiler_fprintf(...) { \
+		profiler_fprintf_pre(); \
+		profiler_fprintf_post(sprintf(XG_PROF(profile_file_buffer) + XG_PROF(profile_file_buffer_len), __VA_ARGS__)); \
+	}
+
+static void profiler_write_header(char *script_name)
 {
 	if (XINI_PROF(profiler_append)) {
-		fprintf(file, "\n==== NEW PROFILING FILE ==============================================\n");
+		profiler_fprintf("\n==== NEW PROFILING FILE ==============================================\n");
```

Comment:
Instead of adding a `profile_fprintf` macro, I would rather see this using:
`xdebug_str_add_literal()`.